### PR TITLE
PYI-715: Authenticate client with client_assertion jwt

### PR DESCRIPTION
**Pull request by dannyhayes-gds**
### What changed
Authenticate client with client_assertion jwt in incoming /token request.

### Why did it change
Client authentication is required for incoming /token requests based off which clients we have configured.

### Issue tracking

- [PYI-715](https://govukverify.atlassian.net/browse/PYI-715)

### CODE REVIEW

**Review by Wynndow — DISMISSED** [review]

**Review by Wynndow — COMMENTED** [review]
I've left a few comments on this PR which I just realised might also apply here:
https://github.com/alphagov/di-ipv-core-back/pull/132

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/AccessTokenHandlerTest.java [hunk 1]
```diff
@@ -190,6 +196,31 @@ void shouldReturn400WhenInvalidRedirectUriParamIsProvided() throws Exception {
...
+    @Test
+    void shouldReturn400WhenClientAuthFails() throws Exception {
+        String tokenRequestBody =
+                "code=12345&redirect_uri=http://test.com&grant_type=authorization_code&client_id=test_client_id";
+
+        APIGatewayProxyRequestEvent event = new APIGatewayProxyRequestEvent();
+        event.setBody(tokenRequestBody);
+
+        ValidationResult<ErrorResponse> validationResultError =
+                new ValidationResult<>(false, ErrorResponse.INVALID_REQUEST_PARAM);
```
Comment (Wynndow) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/AccessTokenHandlerTest.java:208]: I think this can be dropped.

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        OffsetDateTime offsetDateTime =
+                OffsetDateTime.now().plusSeconds(Long.parseLong(maxAllowedTtl));
+        if (expirationTime.getTime() / 1000L > offsetDateTime.toEpochSecond()) {
+            LOGGER.error("Client JWT expiry date is too far in the future");
+            throw new InvalidClientException(
+                    "The client JWT expiry date has surpassed the maximum allowed ttl value");
+        }
+    }
+
+    public static ClientAuthenticationVerifier<Object> getClientAuthVerifier(
```
Comment (Wynndow) [inline lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java:108]: I think this doesn't need to be static as it's being called in the constructor directly.

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                configurationService.getClientAuthenticationMethod(clientId);
+        if (clientAuthenticationMethod == null) {
+            LOGGER.error("Failed to find config for client ID: {}", clientId);
+            throw new ClientAuthenticationException(
+                    String.format("Config for client ID '%s' not found", clientId));
+        }
+        return clientAuthenticationMethod;
+    }
+
+    private void validateExpiration(JWTAuthenticationClaimsSet claimsSet)
```
Comment (Wynndow) [inline lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java:94]: Could this be named to be more descriptive about it validating that the expiry isn't too far in the future?

**Comment by dannyhayes-gds** [comment]
Ah yeah any issues on the other one will probably be relevant here as well. I'll take a look at those and make the same changes here as needed

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+                if (clientAuthenticationMethod.equals(JWT)) {
+                    LOGGER.error("Missing client_assertion jwt for configured client {}", clientId);
+                    throw new ClientAuthenticationException(
+                            String.format(
+                                    "Missing client_assertion jwt for configured client '%s'",
+                                    clientId));
+                }
+            }
+        } else {
```
Comment (Wynndow) [inline lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java:60]: There are quite a few branches in this making it hard to follow - could it be extracted into a well named private method, or broken down in some other way?

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+        StringBuilder sb = new StringBuilder();
+
+        for (Map.Entry<String, String> param : queryParams.entrySet()) {
+            if (sb.length() > 0) {
+                sb.append("&");
+            }
+            sb.append(
+                    String.format(
+                            "%s=%s",
+                            URLEncoder.encode(param.getKey(), "UTF-8"),
```
Comment (Danomanic) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:372]:
```suggestion
                            URLEncoder.encode(param.getKey(), StandardCharsets.UTF_8),
```

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+                            URLEncoder.encode(param.getValue(), "UTF-8")));
```
Comment (Danomanic) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:373]:
```suggestion
                            URLEncoder.encode(param.getValue(), StandardCharsets.UTF_8)));
```

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+                invalidSignatureQueryParams.get("client_assertion") + "BREAKING_THE_SIGNATURE");
+
+        ClientAuthenticationException exception =
+                assertThrows(
+                        ClientAuthenticationException.class,
+                        () -> {
+                            validator.authenticateClient(
+                                    queryMapToString(invalidSignatureQueryParams),
+                                    invalidSignatureQueryParams);
+                            ;
```
Comment (Danomanic) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:103]: Extra semi-colon here :)

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class TokenRequestValidatorTest {
+
+    private TokenRequestValidator validator;
+    @Mock private ConfigurationService mockConfigurationService;
+
+    String clientId = "di-ipv-orchestrator-stub";
```
Comment (Danomanic) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:50]: I know this is a test, but we might want to standardise on `private String` etc.

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+
+    @BeforeEach
+    void setUp() throws CertificateException {
```
Comment (Danomanic) [inline lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:60]: I don't think this throws a `CertificateException` anymore.

File: lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    private static final String CLIENT_ASSERTION_PARAM = "client_assertion";
+    private static final String CLIENT_ID_PARAM = "client_id";
+    private static final String NONE = "none";
+    private static final String JWT = "jwt";
+
+    private final ConfigurationService configurationService;
+
+    private final ClientAuthenticationVerifier<Object> verifier;
+
+    private String clientId;
```
[the remark this answers is not in the document]
Comment (dannyhayes-gds) [reply lambdas/accesstoken/src/main/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidator.java:32]: Ah yeah good catch 👍

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java (hunk 7 above)
Comment (dannyhayes-gds) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:103]: Ooops

File: lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java (hunk 8 above)
Comment (dannyhayes-gds) [reply lambdas/accesstoken/src/test/java/uk/gov/di/ipv/cri/passport/accesstoken/validation/TokenRequestValidatorTest.java:50]: Yeah true, and they can also be final here so I'll change them

**Review by Wynndow — APPROVED** [review]
